### PR TITLE
[MM-409] Add subscription features for project issues and project pull requests

### PR DESCRIPTION
**Pull request by dsspence**
### Summary

This PR is an initial attempt to add a new subscription feature for issues and pull requests assigned to a project https://github.com/mattermost/mattermost-plugin-github/issues/409 .

Since the issue was first proposed, GitHub has introduced a breaking change to their projects implementation creating both a projects (classic) and a projects v2 split. I had to first upgrade the google/go-github dependency to support the newer version of projects https://github.com/mattermost/mattermost-plugin-github/pull/771. When I tried to create an older project resource through the REST API it rejected the request.

Projects V2 are able to be created at the organizational level, user level, and repository level.  This brings into question how much scoping a subscription to a particular repository is to be implemented if projects exist outside of repo boundaries.

I began by trying to re-use the existing `issues` and `pulls` features in the same way as labels are implemented. I quickly decided that this would mean overloading the usage of those features in ways that would make the codebase even more complex such as around duplicate webhook events.  I ultimately decided to segment these into new features titled `project_issues` and `project_pulls`.  There is a new draft issue item as well for projects that doesn't map cleanly to the supported features.

At subscription time this code will use the GraphQL API to find all projects (org and repo level) belonging to the organization and validate that the project title matches one that the user can access (projects can be public or private). This required adding a new oauth scope to the plugin `read:project`, which it didn't have previously.
  
Finally a new webhook event handler was needed to receive the Project V2 Item event, mapping it to an existing project subscription, and building a post action template to send to the Mattermost server.

### Testing Instructions
1. Verify or update your GitHub webhook to include the project v2 items
image
2. Deploy this plugin to your local running Mattermost development server using
```sh
make deploy
```
3. From a channel or private message create a subscription to an organization with an active project
```sh
/github subscriptions add dsspence-org --features project_issues,project:"org project"
```
4. Create a new issue and assign the project used in the previous step from the sidebar before submitting.
5. Observe the new project issue created post in the subscribed Mattermost channel/DM.

### Remaining Steps
- Write automated tests
- Expand GraphQL query to include projects owned by users not just organizations and support pagination beyond first 100 results
- Edge cases such as non-unique project names, trim whitespace, reject empty values ...
- Introduce GraphQL query to lookup issue and pull_requests associated in the Project V2 Item webhook event to provide more detailed post info and/or repo scoping.
- Add pull request template and draft issue support
- Support all of the Project V2 Item event actions beyond just creation (archiving, deleted, restored) and custom templates
- Verify permissions are what we expect in different scenarios

### Open Questions / Observations
- Duplicative default issue subscription is still being shown alongside project issue item subscription
- There was a server setup step around enabling CORS that I would probably add
- A plugin deploy will fail silently if there is a template error, this could use a better error message or additional build step.
- I noticed subscribing with a label only ever uses the first one provided and doesn't support multiple labels in practice. Maybe this should be updated to clarify or support added for checking for all labels in a subscription.
- When attempting to run e2e tests, I think there is an issue with the latest version of Playwright that requires using an older version. I had trouble with the node postinstall rollup command failing for the Mattermost server `webapp/platform/components` package. I had to install dependencies from the child directories because it was failing from the parent folder.
- Node 16 is EOL, are there blockers on moving up to Node 18?

### References
[Organization GraphQL API](https://docs.github.com/en/graphql/reference/queries#organization)
[Projects V2 Item Webhook Event](https://docs.github.com/en/webhooks/webhook-events-and-payloads#projects_v2_item)
[Projects V2 Overview](https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects)

https://github.com/dsspence/mattermost-plugin-github/assets/111698/22450c61-320e-437d-a4b1-8be7a8f98352

### CODE REVIEW

File: server/plugin/subscriptions.go [hunk 1]
```diff
@@ -123,17 +124,44 @@ func (s *Subscription) Stars() bool {
...
 func (s *Subscription) Label() string {
-	if !strings.Contains(s.Features.String(), "label:") {
-		return ""
+	var label string
+	featuresSplit := strings.Split(s.Features.String(), ",")
+
+	for _, feature := range featuresSplit {
+		if strings.Contains(feature, "label:") {
+			labelSplit := strings.Split(feature, "\"")
+			label = labelSplit[1]
```
Comment (dsspence) [inline server/plugin/subscriptions.go:142]: I was surprised to learn you can only subscribe to one label even if it's valid to add multiple in the command features list.

File: server/plugin/subscriptions.go [hunk 2]
```diff
@@ -302,6 +366,26 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
 	return nil
 }
 
+func (p *Plugin) GetSubscribedChannelsForOrg(org string) []*Subscription {
```
Comment (dsspence) [inline server/plugin/subscriptions.go:369]: There is no access to the repository info on the project item event as there are on other webhook events. This would require another API call to scope to what repo the issue or pull request is coming from if the project is at the org level.

**Comment by fmartingr** [comment]
For the reviewers, the command I used to subscribe was:

```
/github subscriptions add <org> --features project_issues,project:"Roadmap"
```

I used `project_issues` instead of `issues`, I think that was a typo. Can you confirm @dsspence ?

**Review by fmartingr — COMMENTED** [review]
Thank you for the Pull Request!

I think you could submit this directly to the main plugin, but I'll follow along with the review here. I really appreciate the effort put into the pull request description, it make me understand and setup my test environment very easily, though I'm afraid to say that I do not receive org notifications myself. Initially it was due to me creating  the webhook using `/github setup webhook`, but I don't receive appropriate events even after setting up it manually. I confirmed that events were being sent ok:

```js
{
  "action": "created",
  "projects_v2_item": {
    // ...
    "content_type": "Issue",
    // ...
}
```

I can't say I approve the PR since I din't make it to work, but I still think there's some missconfiguration on my part, also I'm not very familiar with this project so that's not helping as well. That said, so far what I've seen seems adequate for the task at hand, all considering the limitations of the github webhook payload.

Other observations (probably out of scope of this PR):
- `/github setup webhook` does not setup the appropriate permissions on the webhook. [ref](https://github.com/dsspence/mattermost-plugin-github/blob/master/server/plugin/flows.go#L661)
- We could use raw string interpolation (` `` `) to avoid escaping slashes (`\"`) along the code.
- I receive Issues even if the subscription I create does not subscribe to that. (probably an existing bug? cc: @mickmister)

**Comment by dsspence** [comment]
> /github setup webhook does not setup the appropriate permissions on the webhook. [ref](https://github.com/dsspence/mattermost-plugin-github/blob/master/server/plugin/flows.go#L661)

Ooh yea I followed the repo instructions in the Readme for creating a webhook in GitHub (that's where my Project v2 items screenshot is from) through the UI. You're right we would definitely want to support programmatically creating the webhook through the setup slash command too 👍

File: server/plugin/webhook.go [hunk 3]
```diff
@@ -644,6 +662,84 @@ func (p *Plugin) postIssueEvent(event *github.IssuesEvent) {
...
+		case projectItemActionCreated:
+			contentTemplate = "projectIssueCreated"
+		case projectItemActionDeleted:
+		case projectItemActionEdited:
+		case projectItemActionReordered:
+		case projectItemActionRestored:
+		default:
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (mickmister) [reply server/plugin/webhook.go:694]: I'm having trouble following the current code here. It seems that if `contentTemplate` is blank, we will still call `renderTemplate` below given some other conditions. Does this not result in an invalid template, if we pass an empty string to `renderTemplate`? I could be misreading this though

File: server/plugin/subscriptions.go (hunk 1 above)
Comment (mickmister) [inline server/plugin/subscriptions.go:142]: Is it possible for there to be no `"` character in the `feature` var in this case? That would cause a nil pointer exception on line 142. Same with the `project:` feature. Maybe this is caught by some other validation though

File: server/plugin/subscriptions.go [hunk 4]
```diff
@@ -203,6 +231,42 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
 		Flags:      flags,
 	}
 
+	subscribedProjects := sub.ProjectTitles()
+	if len(subscribedProjects) > 0 {
+		userInfo, apiErr := p.getGitHubUserInfo(userID)
+		if apiErr != nil {
+			return errors.Errorf("Failed to get GitHub user info for userID:%s", userID)
```
Comment (mickmister) [inline server/plugin/subscriptions.go:238]:
```suggestion
			return errors.Wrapf(apiErr, "failed to get GitHub user info for userID:%s", userID)
```

I see we're limiting returning errors elsewhere in this function too, though we are logging them sometimes. Is the goal to avoid revealing information about the errors to the user?

File: server/plugin/subscriptions.go [hunk 5]
```diff
@@ -203,6 +231,42 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
...
+		}
+
+		graphqlClient := p.graphQLConnect(userInfo)
+		if graphqlClient == nil {
+			return errors.Errorf("Failed to get graphql client")
+		}
+
+		// currently only queries for organizational projects not user repo projects
+		// org level projects and repo level projects are both included, both can be used in issue/pr assignment
```
Comment (mickmister) [inline server/plugin/subscriptions.go:247]: Do you mind expanding on this comment? By "organizational projects", is what is meant that org-level and repo-level projects belonging to an org are shown? Then only user-level repo projects are excluded?

```suggestion
		// currently only queries for organizational projects, not user repo projects
		// org level projects and repo level projects are both included, both can be used in issue/pr assignment
```

File: server/plugin/subscriptions.go [hunk 6]
```diff
@@ -203,6 +231,42 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
...
+			var matchedProjects []Project
+			for _, project := range projectData {
+				for _, subscribedProject := range subscribedProjects {
+					if subscribedProject == *project.Title {
+						matchedProjects = append(matchedProjects, Project{NodeID: *project.NodeID, Title: *project.Title})
+					}
+				}
+			}
+			if len(matchedProjects) == 0 {
+				return errors.Errorf("No project(s) were found matching %s", strings.Join(subscribedProjects, ", "))
```
Comment (mickmister) [inline server/plugin/subscriptions.go:264]: Here we return an error if none of the projects exist, but we don't return an error if at least one of the projects doesn't exist. I'm thinking we should probably error out in that case

File: server/plugin/subscriptions.go [hunk 7]
```diff
@@ -302,6 +366,26 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
...
+	subs, err := p.GetSubscriptions()
+	if err != nil {
+		return nil
+	}
+
+	subsForOrg := []*Subscription{}
+	orgKey := org + "/"
+
+	if subs.Repositories[orgKey] != nil {
+		subsForOrg = append(subsForOrg, subs.Repositories[orgKey]...)
```
Comment (mickmister) [inline server/plugin/subscriptions.go:379]: Interesting, so we need the trailing slash to identify a supplied repository as an org in the stored data

File: server/plugin/subscriptions.go [hunk 8]
```diff
@@ -302,6 +366,26 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
 	return nil
 }
 
+func (p *Plugin) GetSubscribedChannelsForOrg(org string) []*Subscription {
+	subs, err := p.GetSubscriptions()
+	if err != nil {
+		return nil
+	}
```
Comment (mickmister) [inline server/plugin/subscriptions.go:373]: We probably want to surface this error, either by logging or returning the error to the caller

File: server/plugin/subscriptions.go [hunk 9]
```diff
@@ -302,6 +366,26 @@ func (p *Plugin) StoreSubscriptions(s *Subscriptions) error {
...
+
+	subsForOrg := []*Subscription{}
+	orgKey := org + "/"
+
+	if subs.Repositories[orgKey] != nil {
+		subsForOrg = append(subsForOrg, subs.Repositories[orgKey]...)
+	}
+
+	if len(subsForOrg) == 0 {
+		return nil
```
Comment (mickmister) [inline server/plugin/subscriptions.go:383]: Returning `nil` here seems error prone. We essentially found zero subs, rather than finding `nil` subs

File: server/plugin/subscriptions.go [hunk 10]
```diff
@@ -203,6 +231,42 @@ func (p *Plugin) Subscribe(ctx context.Context, githubClient *github.Client, use
...
+			p.client.Log.Warn("graphql organization projects query failed", "error", err.Error())
+			return errors.Errorf("Failed to get project data")
+		}
+
+		if projectData != nil {
+			var matchedProjects []Project
+			for _, project := range projectData {
+				for _, subscribedProject := range subscribedProjects {
+					if subscribedProject == *project.Title {
+						matchedProjects = append(matchedProjects, Project{NodeID: *project.NodeID, Title: *project.Title})
```
Comment (mickmister) [inline server/plugin/subscriptions.go:259]: On the subject you brought up regarding having multiple projects with the same name, what's the behavior here when the user retroactively names projects duplicate names after creating the subscription?

File: server/plugin/webhook.go [hunk 11]
```diff
@@ -282,6 +295,11 @@ func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
 		handler = func() {
 			p.postStarEvent(event)
 		}
+	case *github.ProjectV2ItemEvent:
+		// repo info might require another query on a per-event basis if the project is org-based
```
Comment (mickmister) [inline server/plugin/webhook.go:299]: Is this comment essentially a remaining todo?

**Review by mickmister — COMMENTED** [review]
For the purpose of this exercise, definitely approving on my end :+1:  Going with regular "comment" to avoid confusion of it being something we want merged at its current state.

This is excellent work @dsspence. I appreciate the attention to detail and thorough analysis the feature, and other aspects of the codebase. On the review, I have just some comments for discussion. Let me know your thoughts on them. Thank you!

To address some open questions in the PR description:

---

> There was a server setup step around enabling CORS that I would probably add

I'm curious what exactly needed to be done here?

> When attempting to run e2e tests, I think there is an issue with the latest version of Playwright that requires using an older version.

I'm wondering the exact issue you ran into here though? The node version required to run Playwright is lower than the plugin project's node version, or the webapp node version? Note that we are actively working on making e2e testing in plugins more straightforward, and will not require having the monorepo checked out as well.

> Node 16 is EOL, are there blockers on moving up to Node 18?

This is a good point. We're going through a dependency upgrade on all plugin projects, which will include a bump to the node version.

File: server/plugin/webhook.go [hunk 12]
```diff
@@ -644,6 +662,84 @@ func (p *Plugin) postIssueEvent(event *github.IssuesEvent) {
...
+	org := event.GetOrg()
+
+	projectItem := event.GetProjectV2Item()
+	contentType := projectItem.GetContentType()
+	projectNodeID := projectItem.GetProjectNodeID()
+
+	// login matches the graphql org query arg, org name was confusingly empty
+	subscribedChannels := p.GetSubscribedChannelsForOrg(org.GetLogin())
+	if len(subscribedChannels) == 0 {
+		p.client.Log.Debug(fmt.Sprintf("postProjectV2ItemEvent: no subscribed channels org %s", org.GetName()))
```
[the remark this answers is not in the document]
Comment (dsspence) [reply server/plugin/webhook.go:676]: Yep that would make it easier when using something like Cloudwatch

File: server/plugin/subscriptions.go (hunk 10 above)
Comment (dsspence) [reply server/plugin/subscriptions.go:259]: Hmm I'm not sure I specifically did anything to combat that but with more time I would have introduced at least the error on creation if there are duplicate names like we discussed. Successfully created subscriptions should continue to rely on the stable NodeID for matching even if their names change along the way.

**Comment by dsspence** [comment]
@mickmister For CORS, I think I had to go look in the logs for why a particular request was resulting in a time out so it wasn't particularly discoverable.  I might expect it to be called out on a page like this that talks about developing plugins with `ngrok` https://developers.mattermost.com/integrate/plugins/developer-workflow/.

> I'm wondering the exact issue you ran into here though? The node version required to run Playwright is lower than the plugin project's node version, or the webapp node version? Note that we are actively working on making e2e testing in plugins more straightforward, and will not require having the monorepo checked out as well.

The Playwright issue I saw I don't think had to do with the version of Node but that installing a more recent Playwright version (1.42+) would lead to this error https://github.com/microsoft/playwright/issues/29734.

**Comment by raghavaggarwal2308** [comment]
@wiggin77 Can you please have a look at this PR?
